src/Build/Graph/GraphBuilder.cs

```diff
@@ -0,0 +1,397 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Collections;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Experimental.Graph
+{
+    internal class GraphBuilder
+    {
+        private ProjectInterpretation ProjectInterpretation { get; }
+
+        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes { get; private set; }
+
+        public IReadOnlyCollection<ProjectGraphNode> RootNodes { get; private set; }
+
+        public IReadOnlyCollection<ProjectGraphNode> EntryPointNodes { get; private set; }
+
+        public GraphEdges Edges { get; private set; }
+
+        public GraphBuilder(ProjectInterpretation projectInterpretation)
+        {
+            ProjectInterpretation = projectInterpretation;
+        }
+
+        public void BuildGraph(
+            IEnumerable<ProjectGraphEntryPoint> entryPoints,
+            ProjectCollection projectCollection,
+            ProjectGraph.ProjectInstanceFactoryFunc projectInstanceFactory
+            )
+        {
+            var entryPointConfigurationMetadata = AddGraphBuildPropertyToEntryPoints(entryPoints);
+
+            var allParsedProjects = new ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode>();
+            var tasksInProgress = new ConcurrentDictionary<ConfigurationMetadata, object>();
+            var projectsToEvaluate = new ConcurrentQueue<ConfigurationMetadata>();
```

Comment:
Why not just pass the entryPointConfigurationMetadata to the ConcurrentQueue constructor? It is a List and therefore an IEnumerable. If you do that you won't need the foreach below.